sdks/java/io/elasticsearch/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java

```diff
@@ -0,0 +1,303 @@
+      }
+    }
+    return data;
+  }
+  private void insertTestDocuments(long nbDocs) throws Exception {
+    Client client = node.client();
+    final BulkRequestBuilder bulkRequestBuilder = client.prepareBulk().setRefresh(true);
+    List<String> data = createDocuments(nbDocs, false);
+    for (String document:data){
+      bulkRequestBuilder.add(client.prepareIndex(ES_INDEX, ES_TYPE, null).setSource(document));
+    }
+    final BulkResponse bulkResponse = bulkRequestBuilder.execute().actionGet();
+    if (bulkResponse.hasFailures()) {
+      throw new IOException(String.format("Cannot insert test documents in index %s : %s",
+                                          ES_INDEX, bulkResponse.buildFailureMessage()));
+    }
+  }
+
+  @Test
+  public void testSizes() throws Exception {
+    insertTestDocuments(NB_DOCS);
+    PipelineOptions options = PipelineOptionsFactory.create();
+    ElasticsearchIO.Read read =
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration);
+    BoundedElasticsearchSource initialSource =
+        new BoundedElasticsearchSource(read, null);
+    // can't use equal assert as Elasticsearch indexes never have same size
+    // (due to internal Elasticsearch implementation)
+    long estimatedSize = initialSource.getEstimatedSizeBytes(options);
+    LOGGER.info("Estimated size: {}", estimatedSize);
+    assertTrue("Wrong estimated size", estimatedSize > 40000);
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testRead() throws Exception {
+    insertTestDocuments(NB_DOCS);
+
+    TestPipeline pipeline = TestPipeline.create();
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withScrollKeepalive("5m"));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testReadWithQuery() throws Exception {
+    insertTestDocuments(NB_DOCS);
+
+    String query = "{\n"
+        + "  \"query\": {\n"
+        + "  \"match\" : {\n"
+        + "    \"scientist\" : {\n"
+        + "      \"query\" : \"Einstein\",\n"
+        + "      \"type\" : \"boolean\"\n"
+        + "    }\n"
+        + "  }\n"
+        + "  }\n"
+        + "}";
+
+    Pipeline pipeline = TestPipeline.create();
+
+    PCollection<String> output = pipeline.apply(
+        ElasticsearchIO.read().withConnectionConfiguration(connectionConfiguration)
+                              .withQuery(query));
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally())).isEqualTo(NB_DOCS / 10);
+    pipeline.run();
+  }
+
+  @Test
+  @Category(NeedsRunner.class)
+  public void testWrite() throws Exception {
+    Pipeline pipeline = TestPipeline.create();
+    List<String> data = createDocuments(NB_DOCS, false);
```

Comment:
I keep forgetting what the boolean parameter in this method stands for. It's better to replace it with an enum - then the code will read like `createDocuments(NUM_DOCS, DO_NOT_INJECT_INVALID_DOCS or INJECT_SOME_INVALID_DOCS)`